skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,188 @@
+            np.sum((pos/semi_spatial)**2) + (intensity/semi_vertex)**2 = 1
+
+    - This algorithm assums that low intensity values (black) corresponds to
+      the background. If you have a light background, invert the image before
+      passing it into the function, e.g., using `utils.invert`.
+    - This algorithm is sensitive to noise (in particular salt-and-pepper
+      noise). If this is a problem in your image, you can apply mild
+      gaussian smoothing before passing the image to this function.
+
+    Returns
+    -------
+    filtered_image : ndarray
+        The image with background removed.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage import data
+    >>> from skimage.morphology import rolling_ball
+    >>> result = rolling_ball(data.coins())
+
+    References
+    ----------
+    .. [1] Sternberg, Stanley R. "Biomedical image processing." Computer 1
+           (1983): 22-34. :DOI:`10.1109/MC.1983.1654163`
+    """
+
+    kernel_size = np.asarray(kernel_size)
+    if not np.issubdtype(kernel_size.dtype, np.number):
+        raise ValueError(
+            "kernel_size must be convertible to a numeric array.")
+    if not kernel_size.size == 2 and not len(kernel_size.shape) == 1:
+        raise ValueError(
+            "kernel_size must be a two element tuple.")
+    if np.any(kernel_size <= 0):
+        raise ValueError(f"All elements of kernel_size must be greater zero.")
+    kernel_size = kernel_size / 2
+
+    intensity_vertex = np.asarray(intensity_vertex)
+    if not np.issubdtype(intensity_vertex.dtype, np.number):
+        raise ValueError(
+            "Intensity_vertex must be convertible to a numeric array.")
+    if not intensity_vertex.size == 1 and len(intensity_vertex.shape) == 1:
+        raise ValueError(
+            "Intensity_vertex must be a scalar.")
+    if np.any(intensity_vertex <= 0):
+        raise ValueError(f"Intensity_vertex must be greater zero.")
+    intensity_vertex = intensity_vertex / 2
+
+    image = np.asarray(image)
+    if not np.issubdtype(image.dtype, np.number):
+        raise ValueError("Image must be of numeric type.")
+    if not len(image.shape) == 2:
+        raise ValueError("Image must be a two dimensional array.")
+    img = image.copy().astype(np.float_)
+
+    kernel_size_y, kernel_size_x = np.round(kernel_size).astype(int)
+
+    pad_amount = ((kernel_size_x, kernel_size_x),
+                  (kernel_size_y, kernel_size_y))
+    img = np.pad(img, pad_amount,  constant_values=np.Inf, mode="constant")
+
+    tmp_x = np.arange(-kernel_size_x, kernel_size_x + 1)
+    tmp_y = np.arange(-kernel_size_y, kernel_size_y + 1)
+    x, y = np.meshgrid(tmp_x, tmp_y)
+
+    kernel_size_y, kernel_size_x = kernel_size
+    tmp = (x / kernel_size_x) ** 2 + (y / kernel_size_y) ** 2
+    cap_height = intensity_vertex - intensity_vertex * \
+        np.sqrt(np.clip(1 - tmp, 0, None))
+    cap_height = cap_height.astype(np.float_)
```

Comment:
I'm still struggling to completely understand this algorithm:

- So height or intensity of the ellipsoid mean the same?
- If I take the ellipsoid equation
  ```
  (x/a)**2 + (y/b)**2 + (z/c)**2 = 1
  ```
  and solve for `z` I get
  ```
  z(x, y) = | c * sqrt[ 1 - (x/a)**2 - (y/b)**2 ] | 
  ```
  which describes the height of the ellipsoid in the z direction above the xy-plane.
  For those x, y inside the ellipsoid, this corresponds to `intensity_vertex * np.sqrt(np.clip(1 - tmp, 0, None))` right? 
- We are interested in the maximal achievable height of this ellipsoid while it is still covered by the image's "intensity stacks", right? So given the windowed image `w(x, y)` we should be able to compute the background `b` at our current position as
  ```
  b = min_{x, y} w(x, y,) + c - z(x, y)
  ```
  Why the `+ c`? Shouldn't it be `- c` if the background corresponds to the bottom of the ellipsoid or "`+ 0`" if the background corresponds to the center of the ellipsoid? Which of those is true?

Reply:
> So height or intensity of the ellipsoid mean the same?
Yes, I should be consistent with this. Which of the two terms is more intuitive?

>   For those x, y inside the ellipsoid, this corresponds to `intensity_vertex * np.sqrt(np.clip(1 - tmp, 0, None))` right?
Yes.

> * We are interested in the maximal achievable height of this ellipsoid while it is still covered by the image's "intensity stacks", right? 
Exactly.

>   Why the `+ c`? Shouldn't it be `- c` if the background corresponds to the bottom of the ellipsoid or "`+ 0`" if the background corresponds to the center of the ellipsoid? Which of those is true?
Neither of the last two is true; the background is at the apex/top of the ellipsis. The way I think about it is the following: 

To estimate the intensity of the background for some pixel, we introduce a temporary coordinate system that has its origin at that pixel position (the window). In other words, we always estimate the intensity for position `(0,0)` (wlog. ;) ). In this temporary coordinate system, we can easily describe the neighborhood of the pixel as well as the ellipsoid.

If we simplify the algorithm and only consider a rolling disk (the approximation from #3538 ) for a second then the background is simply:
```
b = min_{x,y} w(x,y)
```
where `w(x,y)` is the intensity at that window position.

When we extend this to an ellipsoid (i.e., put a cap on top of the disk), we need to introduce a term that corrects/accounts for the cap. How much higher is the cap in the center (where we estimate the intensity) compared to the position `(x,y)` (where the ellipsoid touches)? At `(0,0)` the height is exactly `c`, at `(x, y)` the height is `z(x,y)`, and the difference `c - z(x.y)` is exactly the amount by which the center is higher than the position `(x,y)`. Hence the background intensity is
```
b = min_{x, y} w(x, y,) + (c - z(x, y))
```

Does that make sense?

Reply:
Ah, yes I think I get it now. I was almost there but didn't consider that it makes more sense to derive background from the ellipsoids top / highest position which is `+ c` from the ellipsoids center. Thanks for taking the time to explain.